spec/asyncapi.md

```diff
@@ -1426,6 +1434,14 @@ description: Find more info here
 url: https://example.com
 ```
 
+#### <a name="schemaReferenceLinkingObject"></a>Schema Reference and Linking Object
+
+The Schema Reference and Linking Object is defined by [draft-asyncapi-ref-linking-01](./draft-asyncapi-ref-linking-01.md).
+
+The difference between this and normal [Reference Object](#referenceObject), is to support linking to non-JSON data, that is otherwise not possible through the Reference Object. As well as defining a behavior for when encountering subsequent references.
+
+See [schema format](....) for reference-linking formats that is supported and how it's defined.
+
```

Comment:
@jonaslagoni thanks for working through all of this!

> do you @handrews think that JSON Reference properly align with the requirements (as listed in the PR description)?

So, I realize that my first big comment goes along with JSON Reference as the starting point, and then I came back later and said "don't do that, it's been abandoned for nearly a decade" - sorry about the mixed messaging.  I do think it's better to treat the long-abandoned JSON Reference as an inspiration at most, and not something that can or should be normatively cited.

JSON Pointer (RFC 6901) can be normatively cited, either as "we are using RFC 6901 directly" or "we are using RFC 6901 with the following extensions."  For example, JSON Schema at times uses JSON Pointers (not URI fragments, just JSON Pointers) with instance data, which JSON works with through a _data model_ that is defined based on JSON.  Which is really what happens whenever you parse JSON and then apply a JSON Pointer:  you're not working on the document, you're working on an in-memory representation, which is how most people work with JSON Pointers anyway.  It's just such a trivial mapping that no one mentions it.

This brings us to:

>  Do you see it as a problem that we explicitly define the media type for a reference instead of having it determined when the resource of the URI is resolved?... during the resolution phase of the `http://example.com/example.json` determine the mime-type to be `application/schema+json`, how can define such contradicting behaviours then 🤔

So, you get a resource and you know it's media type (and I'm guessing with some of these formats, there might not be a proper media type defined for them, but you have some other way of knowing what the format is).  Technically, (as I understand it anyway), the correct thing to do is interpret the fragment based on the media type, and if the media type does not allow fragments, or doesn't understand that fragment's syntax, then you just end up with the whole resource.

(this next bit might contradict earlier stuff I said about mapping into a different media type, but this is probably a more solid approach)

JSON Hyper-Schema had this problem, because it was designed to work with `application/json` (no fragment syntax defined) and potentially other media types that were even less JSON Pointer-fragment-friendly that we couldn't anticipate.

Whenever we needed to identify a resource by a URI, and then access some part of it that would conveniently be accessed by a JSON Pointer fragment, we instead used a URI that was _not_ allowed to have a fragment, and paired it with a plain JSON Pointer (not encoded in a URI).  This is why [`anchorPointer`](https://datatracker.ietf.org/doc/html/draft-handrews-json-schema-hyperschema#section-6.1.2) is a separate field from [`anchor`](https://datatracker.ietf.org/doc/html/draft-handrews-json-schema-hyperschema#section-6.1.1) (these URIs apply to the instance).

Actually, I think we don't forbid fragments in hyper-schema's `anchor` to leave people the flexibility of either using a fragment syntax if they know it will work, or using a non-fragment JSON Pointer if they aren't sure but know that they'll be able to figure out how to apply the JSON Pointer.

While plain JSON Pointers are also officially defined on JSON _documents_, JSON Schema works on a _data model_ that is based on JSON so as long as the instance, regardless of its media type, has been mapped into the data model, we can apply a JSON Pointer.  Because, really, most code applies JSON Pointers to in-memory parsed JSON data, not the actual document.  So we just say "we're using an in-memory structure that looks like what a JSON parser produces, and we apply JSON Pointer to that."

It's less elegant than URI fragments, _but_ it avoids all of the pitfalls of the media-type-defined nature of URI fragments.  (We do use JSON Pointer fragments _and_ plain name fragments in JSON Schema's `$ref` because it is always pointing to a JSON Schema, and JSON Schema's media type defines both JSON Pointer and plain name fragments).

> Do you also see that we need to define the behaviour for how you map (for example) XML into JSON structure once resolved? I.e. currently I am just lazy and resolve it into a string and let tooling decide how they want to represent and access it.

If you go with the non-fragment JSON Pointer approach, then yes you need to define how to map XML (or wahtever) into the JSON data model, so it ends up looking like it could have been parsed from JSON.  For some formats that's trivial, for others (including XML) it's ambiguous and you have to make some decisions about what to do.

You can specify that for each significant format, though, and I think a lot of the ones you are using are pretty straightforward to map.

(out of time for now, I'll come back and look at more later, but feel free to ping me if I forget).

Reply:
> This is precisely why it might be tough to reuse some OpenAPI Schema objects in AsyncAPI because you might have tied the references up with a base URI defined in OpenAPI because when we then reference the OpenAPI Schema object, that base URI is no longer accessable.
> 
> But then again, the same issue might happen with JSON Schema documents I guess... Not really anything we can do about it I guess. We just simply have to be aware it might be a problem for some.

I'm not entirely sure I follow.  How is the base URI no longer accessible?  If you accessed the OAS document, then shouldn't you be able to read the base URI from it?  And if you are accessing JSON Schema _documents_, those have the standard JSON Schema base URI rules.

If you are handed a Schema Object from an OAS file and do not have access to the OAS file, then I can see how that would be a problem, but this is why for any JSON Schema you want to re-use across multiple contexts, you should set an absolute URI for `$id` to ensure that the base URI is defined regardless of context.  That's what `$id` is for.